File: src/main/scala/com/diyishuai/algorithm/scala/BinaryTreeNode.scala
package com.diyishuai.algorithm.scala


case class BinaryTreeNode(leftChild: Option[BinaryTreeNode], value: Long, rightChild: Option[BinaryTreeNode]) {

}

object BinaryTreeNode {

  def preOrder(treeNode: Option[BinaryTreeNode]): Unit = {
    treeNode match {
      case Some(node) => {
        println(node.value)
        preOrder(node.leftChild)
        preOrder(node.rightChild)
      }
      case None => {}
    }
  }

  def inOrder(treeNode: Option[BinaryTreeNode]): Unit = {
    treeNode match {
      case Some(node) => {
        inOrder(node.leftChild)
        println(node.value)
        inOrder(node.rightChild)
      }
      case None => {}
    }
  }


  def postOrder(treeNode: Option[BinaryTreeNode]): Unit = {
    treeNode match {
      case Some(node) => {
        postOrder(node.leftChild)
        postOrder(node.rightChild)
        println(node.value)
      }
      case None => {}
    }
  }

  def main(args: Array[String]): Unit = {
    /**
      *    1
      *  2   3
      * 4   5 6
      */
    val root = new BinaryTreeNode(
      Some(new BinaryTreeNode(
        Some(new BinaryTreeNode(
          None,
          4l,
          None
        )),
        2l,
        None
      )),
      1l,
      Some(new BinaryTreeNode(
        Some(new BinaryTreeNode(
          None,
          5l,
          None
        )),
        3l,
        Some(new BinaryTreeNode(
          None,
          6l,
          None
        ))
      ))
    )

    preOrder(Some(root))
    println("====")
    inOrder(Some(root))
    println("====")
    postOrder(Some(root))
  }

}

File: src/main/java/com/diyishuai/algorithm/java/BinaryTreeNode.java
package com.diyishuai.algorithm.java;

/**
 * @author Bruce
 * @since 2019-04-10
 */
public class BinaryTreeNode {

    public BinaryTreeNode leftChild;

    public Long value;

    public BinaryTreeNode rightChild;



    public BinaryTreeNode(BinaryTreeNode leftChild, Long value, BinaryTreeNode rightChild) {
        this.leftChild = leftChild;
        this.rightChild = rightChild;
        this.value = value;
    }

    public BinaryTreeNode() {
    }

    public static void preOrder(BinaryTreeNode treeNode){
        if (treeNode!=null){
            System.out.println(treeNode.value);
            preOrder(treeNode.leftChild);
            preOrder(treeNode.rightChild);
        }
    }

    public static void inOrder(BinaryTreeNode treeNode){
        if (treeNode!=null){
            inOrder(treeNode.leftChild);
            System.out.println(treeNode.value);
            inOrder(treeNode.rightChild);
        }
    }

    public static void postOrder(BinaryTreeNode treeNode){
        if (treeNode!=null){
            postOrder(treeNode.leftChild);
            postOrder(treeNode.rightChild);
            System.out.println(treeNode.value);
        }
    }


    public static void main(String[] args) {
        /**
         *               1
         *             2   3
         *            4   5 6
         */
        BinaryTreeNode root = new BinaryTreeNode(
                new BinaryTreeNode(
                        new BinaryTreeNode(null,4l,null),
                        2L,
                        null
                ),
                1L,
                new BinaryTreeNode(
                        new BinaryTreeNode(null,5l,null),
                        3L,
                        new BinaryTreeNode(null,6l,null)
                )
        );

        preOrder(root);
        System.out.println("======");
        inOrder(root);
        System.out.println("======");
        postOrder(root);
    }
}
